### ISSUE TITLE
Vectorized logical functions

### ISSUE THREAD

**Issue by jessexknight** [stan-dev/math#2925]
## Description
As discussed [here](https://github.com/stan-dev/math/issues/202#issuecomment-1638751377), the [logical functions](https://mc-stan.org/docs/functions-reference/logical-functions.html) don't support vectorization yet. I think this should return N-sized outputs for both: a) two N-sized inputs (element-wise comparison), and b) one N-sized input and one scalar input (broadcasting). As @bob-carpenter noted, we will likely want to add `any()` and `all()` functions to process these outputs.

## Example & Expected Output
R:
```R
N = 100
stan(
  file='tmp.stan',
  data=list(N=N,
            x1=runif(N),
            x2=runif(N)))
```
stan:
```stan
data {
  int N;
  real x1[N];
  real x2[N];
}
transformed data {
  real y[N];
  y = sum(x1 > x2); // desired
  // for (i in 1:N){ // required alternative (slow)
  //   y[i] += x1[i] > x2[i];
  // }
}
```
Current error:
```
No matches for: logical_gt(real[ ], real[ ])
```
This code is invalid for other reasons, but it should at least compile past the current error.

#### Current Version:
v4.6.2

**Comment by bob-carpenter** [comment]
Thanks for opening the issue, @jessexknight.  I edited to add signatures and a definition and to remove the unnecessary R and remove the erroneous comment about efficiency.

As far as efficiency, I'm afraid implementing a built-in in C++ won't be any faster than writing the loop in Stan, because Stan gets compiled down to C++.  The only advantage to having built-ins is when we have vectorized autodiff, which we can accelerate.

**Comment by jessexknight** [comment]
Thanks - that is surprising about performance. I'm really out of my depth here with C++ etc., but would there be a performance difference if using something like `apply_scalar_binary` to existing scalar functions, vs a native function for this kind of thing from the _Eigen_ library?

**Comment by bob-carpenter** [comment]
Yes, if we can get things compiled down to Eigen's vectorized operations then we can exploit their use of CPU vectorization (e.g., SSE and AVX operations at the CPU level).  This can give a several times speedup.  The term "vectorized" usually refers to using SSE, AVX, etc. on the CPU---our use of the term in Stan is non-standard.  I doubt they've vectorized logical operations, but they've done a lot of common math functions like log and exp and sin and cos.

Vectorized CPU operations can be a lot faster.  

`apply_scalar_binary` doesn't compile down that low but we might actually be able to rewrite it to better exploit these operations.  I think the current implementation assumes a double or autodiff.
